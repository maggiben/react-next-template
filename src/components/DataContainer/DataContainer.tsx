import styled, { css } from "styled-components";
import { TextBody } from '@fravega-it/bumeran-ds-fvg'
import { splitIntoTuples } from '@utils/array';
import { SpaceRight } from '@components/Spacing/Spacing';

const getColumnWidth = (index: number, width: number) => {
  return css`
  th:nth-child(${index + 1}),
  td:nth-child(${index + 1}) {
      width: ${width}%;
    }
  `;
};

const calculateColumns = (columns: number) => 
  Array.from({ length: columns}, (_, index) => getColumnWidth(index, 100 / columns)).join(String.fromCharCode(10));

const Table = styled.table<{columns: number, withBorder?: boolean, background?: string}>`
  width: 100%;
  ${props => props.withBorder && css`
    border-width: 1px;
    border-color: ${({ theme }) => theme.colors.neutral[200]};
    border-style: solid;
    border-radius: ${({ theme }) => theme.borderRadius.s};
  `}
  th, td {
    padding: 8px;
    text-align: left;
  }

  /* Set the width of each column */
  ${({ columns }) => calculateColumns(columns)}
  background-color: ${({ theme }) => theme.colors.neutral[100]};
  ${props => props.background && css`
    background-color: ${props.background};
  `}
`;

const Centered = styled.div`
  display: flex;
  justify-content: start;
  align-items: center;
`;

export interface IDataContainerProps {
  columns: number;
  withBorder?: boolean;
  data: Array<Record<string, string>>;
  background?: string;
}

const DataContainer = ({ data, columns, withBorder, background}: IDataContainerProps) => {
  const tuples = splitIntoTuples(data, columns);
  return (
    <Table columns={columns} withBorder={withBorder} background={background}>
      <tbody>
        {tuples.map((tuple, index) => (
          <tr key={index}>
            {tuple.map((value, columnIndex) => {
              return (
                <td key={columnIndex}>
                  {
                    Object.entries(value).map(([label, description]) => (
                      <Centered key={label}>
                        <TextBody size="m" color="neutral" as="span">{label}</TextBody>
                        <SpaceRight size="s" />
                        <TextBody size="m" as="span">{description}</TextBody>
                      </Centered>
                    ))
                  }
                </td>
              );
            })
          }
          </tr>
        ))}
      </tbody>
    </Table>
  );
};

export default DataContainer;