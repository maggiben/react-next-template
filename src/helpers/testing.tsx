import { default as baseRenderer } from 'react-test-renderer';
import { render } from '@testing-library/react';
import i18n from '@utils/i18n';
import { I18nextProvider } from 'react-i18next';
import { ThemeProvider } from 'styled-components';
import { theme, IDefaultTheme } from '@fravega-it/bumeran-ds-fvg';
import React, { ReactElement } from 'react';

/**
 * Helper used to render components for testing.
 * The function appends the theme to the children.
 * */
interface IRenderer {
  create: (children: ReactElement) => any;
  baseRenderer: (children: ReactElement) => any;
  renderWithTheme: (children: ReactElement, theme: IDefaultTheme) => any;
  renderWithI18n: (children: ReactElement) => any;
}

export const renderer: IRenderer = {
  create: (children: ReactElement) =>
    render(
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
  ),
  renderWithTheme: (children: ReactElement, theme: IDefaultTheme) =>
    render(
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
  ),
  baseRenderer: (children: ReactElement) =>
    baseRenderer.create(
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
  ),
  renderWithI18n: (children: ReactElement) => 
    render(
      <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
  ),
};