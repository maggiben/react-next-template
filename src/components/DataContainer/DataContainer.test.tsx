import React from 'react';
import { renderer } from '@helpers/testing';
import '@testing-library/jest-dom';
import {
  screen,
  getAllByRole,
} from "@testing-library/react";
import DataContainer from './DataContainer';
import { expect } from '@jest/globals';

// Mock the utility function used in the component
// jest.mock('@utils/array', () => ({
//   splitIntoTuples: jest.fn(data => data), // Mock the splitIntoTuples function to return the input data
// }));

describe('DataContainer', () => {
  const data = [
    {
      label: 'value',
    }, {
      label: 'value',
    }, {
      label: 'value',
    }, {
      label: 'value',
    },{
      label: 'value',
    },{
      label: 'value',
    },{
      label: 'value',
    },{
      label: 'value',
    },
  ];

  it('renders the DataContainer component', () => {
    const tree = renderer.create(<DataContainer data={data} columns={2} />);
    expect(tree).toMatchSnapshot();
    // expect(tree).toBeInTheDocument();
  });

  it('renders the correct number of rows and columns', () => {
    const { getAllByRole } = renderer.create(<DataContainer data={data} columns={2} />);
    // Query all cells using the 'cell' role
    const cells = getAllByRole('cell');

    // Check if the correct number of cells is rendered
    expect(cells).toHaveLength(data.length);

    // You can also perform additional assertions on the cells if needed
    cells.forEach((cell: HTMLTableCellElement) => {
      expect(cell).toHaveTextContent(/labelvalue/ig); // For example, check if each cell contains text matching the pattern "value1", "value2", etc.
    });
  });
});