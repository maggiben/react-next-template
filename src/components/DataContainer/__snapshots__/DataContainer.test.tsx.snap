// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataContainer renders the DataContainer component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="DataContainer__Table-sc-122pn0e-0 dPZjvX"
      >
        <tbody>
          <tr>
            <td>
              <div
                class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
              >
                <p
                  class="sc-gsFSXq btaipV"
                  color="neutral"
                >
                  label
                </p>
                <div
                  class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
                />
                <p
                  class="sc-gsFSXq fJmkby"
                >
                  value
                </p>
              </div>
            </td>
            <td>
              <div
                class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
              >
                <p
                  class="sc-gsFSXq btaipV"
                  color="neutral"
                >
                  label
                </p>
                <div
                  class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
                />
                <p
                  class="sc-gsFSXq fJmkby"
                >
                  value
                </p>
              </div>
            </td>
            <td>
              <div
                class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
              >
                <p
                  class="sc-gsFSXq btaipV"
                  color="neutral"
                >
                  label
                </p>
                <div
                  class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
                />
                <p
                  class="sc-gsFSXq fJmkby"
                >
                  value
                </p>
              </div>
            </td>
            <td>
              <div
                class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
              >
                <p
                  class="sc-gsFSXq btaipV"
                  color="neutral"
                >
                  label
                </p>
                <div
                  class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
                />
                <p
                  class="sc-gsFSXq fJmkby"
                >
                  value
                </p>
              </div>
            </td>
          </tr>
          <tr>
            <td>
              <div
                class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
              >
                <p
                  class="sc-gsFSXq btaipV"
                  color="neutral"
                >
                  label
                </p>
                <div
                  class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
                />
                <p
                  class="sc-gsFSXq fJmkby"
                >
                  value
                </p>
              </div>
            </td>
            <td>
              <div
                class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
              >
                <p
                  class="sc-gsFSXq btaipV"
                  color="neutral"
                >
                  label
                </p>
                <div
                  class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
                />
                <p
                  class="sc-gsFSXq fJmkby"
                >
                  value
                </p>
              </div>
            </td>
            <td>
              <div
                class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
              >
                <p
                  class="sc-gsFSXq btaipV"
                  color="neutral"
                >
                  label
                </p>
                <div
                  class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
                />
                <p
                  class="sc-gsFSXq fJmkby"
                >
                  value
                </p>
              </div>
            </td>
            <td>
              <div
                class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
              >
                <p
                  class="sc-gsFSXq btaipV"
                  color="neutral"
                >
                  label
                </p>
                <div
                  class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
                />
                <p
                  class="sc-gsFSXq fJmkby"
                >
                  value
                </p>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="DataContainer__Table-sc-122pn0e-0 dPZjvX"
    >
      <tbody>
        <tr>
          <td>
            <div
              class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
            >
              <p
                class="sc-gsFSXq btaipV"
                color="neutral"
              >
                label
              </p>
              <div
                class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
              />
              <p
                class="sc-gsFSXq fJmkby"
              >
                value
              </p>
            </div>
          </td>
          <td>
            <div
              class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
            >
              <p
                class="sc-gsFSXq btaipV"
                color="neutral"
              >
                label
              </p>
              <div
                class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
              />
              <p
                class="sc-gsFSXq fJmkby"
              >
                value
              </p>
            </div>
          </td>
          <td>
            <div
              class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
            >
              <p
                class="sc-gsFSXq btaipV"
                color="neutral"
              >
                label
              </p>
              <div
                class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
              />
              <p
                class="sc-gsFSXq fJmkby"
              >
                value
              </p>
            </div>
          </td>
          <td>
            <div
              class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
            >
              <p
                class="sc-gsFSXq btaipV"
                color="neutral"
              >
                label
              </p>
              <div
                class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
              />
              <p
                class="sc-gsFSXq fJmkby"
              >
                value
              </p>
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div
              class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
            >
              <p
                class="sc-gsFSXq btaipV"
                color="neutral"
              >
                label
              </p>
              <div
                class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
              />
              <p
                class="sc-gsFSXq fJmkby"
              >
                value
              </p>
            </div>
          </td>
          <td>
            <div
              class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
            >
              <p
                class="sc-gsFSXq btaipV"
                color="neutral"
              >
                label
              </p>
              <div
                class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
              />
              <p
                class="sc-gsFSXq fJmkby"
              >
                value
              </p>
            </div>
          </td>
          <td>
            <div
              class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
            >
              <p
                class="sc-gsFSXq btaipV"
                color="neutral"
              >
                label
              </p>
              <div
                class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
              />
              <p
                class="sc-gsFSXq fJmkby"
              >
                value
              </p>
            </div>
          </td>
          <td>
            <div
              class="DataContainer__Centered-sc-122pn0e-1 byuuFh"
            >
              <p
                class="sc-gsFSXq btaipV"
                color="neutral"
              >
                label
              </p>
              <div
                class="Spacing__SpaceRight-sc-11df7di-0 latkgH"
              />
              <p
                class="sc-gsFSXq fJmkby"
              >
                value
              </p>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
